public/static/app/perf/TrendSamples.js

```diff
@@ -0,0 +1,131 @@
+    this.samples = samples;
+    var NON_THREAD_LEVELS = ['start', 'end']
+
+    // _sampleByCommitIndexes is a map of mappings of (githash -> sample data), keyed by test name.
+    // e.g.
+    // {
+    //   "test-foo":  {"ab215e..." : { sample data }, "bced3f..." : { sample data }, ...
+    //   "test-blah": {"ab215e..." : { sample data }, "bced3f..." : { sample data }, ...
+    //   ..
+    //}
+    this._sampleByCommitIndexes = {};
+
+    // seriesByName is a mapping of test names to sample data.
+    this.seriesByName = {};
+
+    this._tasksByName = {}
+
+    // testNames is a unique list of all the tests that appear in *any* of the given list of samples.
+    this.testNames = [];
+
+    // Available metrics (ops_per_sec is default and mandatory one)
+    this.metrics = []
+
+    // Internal-use metrics accumilator hash table
+    const metricsSet = new Set()
+
+    for (var i = 0; i < samples.length; i++) {
+      var sample = samples[i];
+
+      for (var j = 0; j < sample.data.results.length; j++) {
+        var rec = sample.data.results[j];
+
+        // Create entry if not exists
+        if (!(rec.name in this.seriesByName)) {
+          this.seriesByName[rec.name] = [];
+        }
+
+        var maxThreadOpsPerSecItem = _.max(rec.results, function(d) {
+          return d.ops_per_sec
+        })
+
+        // Sort items by thread level
+        // Change dict to array
+        var threadResults = _.chain(rec.results)
+          .omit(NON_THREAD_LEVELS)
+          .map(function(v, k) {
+            // TODO will remove this soon
+            //if (i > 50) {
+            //  v.other_metric = v.ops_per_sec * (.5 + 3 * Math.random())
+            //  v.other_metric_values = [v.other_metric]
+            //}
+            _.each(_.keys(v), (d) => metricsSet.add(d))
+            v.threadLevel = k
+            return v
+          })
+          .sortBy('-threadLevel')
+          .value()
+
+        this.seriesByName[rec.name].push({
+          revision: sample.revision,
+          task_id: sample.task_id,
+          ops_per_sec: maxThreadOpsPerSecItem.ops_per_sec,
+          ops_per_sec_values: maxThreadOpsPerSecItem.ops_per_sec_values,
+          order: sample.order,
+          createTime: sample.create_time,
+          threadResults: threadResults,
+        });
+      }
+    }
+
+    // copy unique items to an array
+    this.metrics = _.reject([...metricsSet], (d) => d.endsWith('_values'))
+    // free mem
+    metricsSet.clear()
+
+    for (let key in this.seriesByName) {
+      this.seriesByName[key] = _.sortBy(this.seriesByName[key], 'order');
+      this.testNames.unshift(key);
+    }
+
+    for (let i = 0; i < this.testNames.length; i++) {
+      //make an index for commit hash -> sample for each test series
+      var k = this.testNames[i];
+      // FIXME Unknown behavior (coma operator after _groupBy stmt)
+      this._sampleByCommitIndexes[k] = _.groupBy(this.seriesByName[k], "revision"), function(x){return x[0]};
```

Comment:
I think the closing parens is in the wrong place here

Reply:
Right. That's why I left a comment. I have no idea how this supposed to work. And no one will tell you, I think. This code was copied several times before I found this potential bug with SonarQube.